File: ta4j-core/src/main/java/org/ta4j/core/analysis/CashFlow.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.analysis;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import java.util.Objects;
import java.util.stream.Stream;
import org.ta4j.core.BarSeries;
import org.ta4j.core.BaseTradingRecord;
import org.ta4j.core.Position;
import org.ta4j.core.Trade;
import org.ta4j.core.TradingRecord;
import org.ta4j.core.num.Num;

/**
 * Allows to follow the money cash flow involved by a list of positions over a
 * bar series, either marked to market or using realized values only.
 */
public class CashFlow implements PerformanceIndicator {

    /**
     * The bar series.
     */
    private final BarSeries barSeries;

    /**
     * The (accrued) cash flow sequence (without trading costs).
     */
    private final OffsetNumBuffer values;

    /**
     * The window captured when the cash flow was materialized.
     */
    private final AnalysisPositionSupport.Window window;

    /**
     * The equity curve calculation mode.
     */
    private final EquityCurveMode equityCurveMode;

    /**
     * Constructor.
     *
     * @param barSeries            the bar series
     * @param tradingRecord        the trading record
     * @param finalIndex           index up until cash flows of open positions are
     *                             considered
     * @param equityCurveMode      the calculation mode
     * @param openPositionHandling how to handle open positions
     * @since 0.22.2
     */
    public CashFlow(BarSeries barSeries, TradingRecord tradingRecord, int finalIndex, EquityCurveMode equityCurveMode,
            OpenPositionHandling openPositionHandling) {
        this(barSeries, tradingRecord, 0, finalIndex, equityCurveMode, openPositionHandling, false, false, true);
    }

    /**
     * Constructor materializing only a bounded logical window on the original
     * series.
     *
     * @param barSeries            the bar series
     * @param tradingRecord        the trading record
     * @param startIndex           first absolute bar index to materialize
     * @param finalIndex           last absolute bar index to materialize and to
     *                             consider for open positions
     * @param equityCurveMode      the calculation mode
     * @param openPositionHandling how to handle open positions
     * @since 0.22.5
     */
    public CashFlow(BarSeries barSeries, TradingRecord tradingRecord, int startIndex, int finalIndex,
            EquityCurveMode equityCurveMode, OpenPositionHandling openPositionHandling) {
        this(barSeries, tradingRecord, startIndex, finalIndex, equityCurveMode, openPositionHandling, false, false,
                false);
    }

    /**
     * Constructor for cash flows of a closed position.
     *
     * @param barSeries       the bar series
     * @param position        a single position
     * @param equityCurveMode the calculation mode
     * @since 0.22.2
     */
    public CashFlow(BarSeries barSeries, Position position, EquityCurveMode equityCurveMode) {
        this(barSeries, new BaseTradingRecord(position), 0, 0, equityCurveMode, OpenPositionHandling.MARK_TO_MARKET,
                false, true, true);
    }

    /**
     * Constructor.
     *
     * @param barSeries       the bar series
     * @param tradingRecord   the trading record
     * @param finalIndex      index up until cash flows of open positions are
     *                        considered
     * @param equityCurveMode the calculation mode
     * @since 0.22.2
     */
    public CashFlow(BarSeries barSeries, TradingRecord tradingRecord, int finalIndex, EquityCurveMode equityCurveMode) {
        this(barSeries, tradingRecord, finalIndex, equityCurveMode, OpenPositionHandling.MARK_TO_MARKET);
    }

    /**
     * Constructor for cash flows of a closed position.
     *
     * @param barSeries the bar series
     * @param position  a single position
     */
    public CashFlow(BarSeries barSeries, Position position) {
        this(barSeries, position, EquityCurveMode.MARK_TO_MARKET);
    }

    /**
     * Constructor for cash flows of closed positions of a trading record.
     *
     * @param barSeries     the bar series
     * @param tradingRecord the trading record
     */
    public CashFlow(BarSeries barSeries, TradingRecord tradingRecord) {
        this(barSeries, tradingRecord, 0, 0, EquityCurveMode.MARK_TO_MARKET, OpenPositionHandling.MARK_TO_MARKET, true,
                false, true);
    }

    /**
     * Constructor.
     *
     * @param barSeries       the bar series
     * @param tradingRecord   the trading record
     * @param equityCurveMode the calculation mode
     * @since 0.22.2
     */
    public CashFlow(BarSeries barSeries, TradingRecord tradingRecord, EquityCurveMode equityCurveMode) {
        this(barSeries, tradingRecord, 0, 0, equityCurveMode, OpenPositionHandling.MARK_TO_MARKET, true, false, true);
    }

    /**
     * Constructor.
     *
     * @param barSeries            the bar series
     * @param tradingRecord        the trading record
     * @param equityCurveMode      the calculation mode
     * @param openPositionHandling how to handle open positions
     * @since 0.22.2
     */
    public CashFlow(BarSeries barSeries, TradingRecord tradingRecord, EquityCurveMode equityCurveMode,
            OpenPositionHandling openPositionHandling) {
        this(barSeries, tradingRecord, 0, 0, equityCurveMode, openPositionHandling, true, false, true);
    }

    /**
     * Constructor.
     *
     * @param barSeries     the bar series
     * @param tradingRecord the trading record
     * @param finalIndex    index up until cash flows of open positions are
     *                      considered
     */
    public CashFlow(BarSeries barSeries, TradingRecord tradingRecord, int finalIndex) {
        this(barSeries, tradingRecord, finalIndex, EquityCurveMode.MARK_TO_MARKET, OpenPositionHandling.MARK_TO_MARKET);
    }

    /**
     * Constructor.
     *
     * @param barSeries            the bar series
     * @param tradingRecord        the trading record
     * @param openPositionHandling how to handle open positions
     * @since 0.22.2
     */
    public CashFlow(BarSeries barSeries, TradingRecord tradingRecord, OpenPositionHandling openPositionHandling) {
        this(barSeries, tradingRecord, 0, 0, EquityCurveMode.MARK_TO_MARKET, openPositionHandling, true, false, true);
    }

    private CashFlow(BarSeries barSeries, TradingRecord tradingRecord, int startIndex, int requestedFinalIndex,
            EquityCurveMode equityCurveMode, OpenPositionHandling openPositionHandling, boolean useRecordEnd,
            boolean useSeriesEnd, boolean padToSeriesEnd) {
        this.barSeries = Objects.requireNonNull(barSeries, "barSeries");
        this.equityCurveMode = Objects.requireNonNull(equityCurveMode);
        TradingRecord record = Objects.requireNonNull(tradingRecord);
        OpenPositionHandling handling = Objects.requireNonNull(openPositionHandling);
        AnalysisPositionSupport.Curve curve = barSeries.withReadLock(() -> {
            AnalysisPositionSupport.Window captured = AnalysisPositionSupport.captureWindow(this.barSeries, record,
                    startIndex, requestedFinalIndex, useRecordEnd, useSeriesEnd, padToSeriesEnd);
            Num one = this.barSeries.numFactory().one();
            OffsetNumBuffer buffer = AnalysisPositionSupport.buffer(captured, one, one);
            for (Position position : AnalysisPositionSupport.positionsForAnalysis(record, captured.finalIndex(),
                    handling, this.equityCurveMode)) {
                calculatePosition(position, captured.finalIndex(), captured, buffer);
            }
            return new AnalysisPositionSupport.Curve(captured, buffer);
        });
        this.window = curve.window();
        this.values = curve.values();
    }

    /**
     * Calculates the cash flow for a single position (including accrued cashflow
     * for open positions).
     *
     * @param position   a single position
     * @param finalIndex index up until cash flow of open positions is considered
     * @since 0.22.2
     */
    @Override
    public void calculatePosition(Position position, int finalIndex) {
        calculatePosition(position, finalIndex, window, values);
    }

    private void calculatePosition(Position position, int finalIndex, AnalysisPositionSupport.Window captured,
            OffsetNumBuffer buffer) {
        Trade entry = position.getEntry();
        if (entry == null) {
            return;
        }
        int addressableEndIndex = captured.addressableEndIndex();
        int entryIndex = entry.getIndex();
        if (entryIndex > finalIndex || entryIndex > addressableEndIndex) {
            return;
        }
        int endIndex = determineEndIndex(position, finalIndex, addressableEndIndex);
        int windowStartIndex = captured.beginIndex();
        int windowEndIndex = captured.bufferEndIndex();
        if (windowStartIndex > windowEndIndex || endIndex < windowStartIndex) {
            return;
        }

        boolean isLongTrade = entry.isBuy();
        Num netEntryPrice = entry.getNetPrice();
        Num entryEquity = buffer.get(Math.max(entryIndex, windowStartIndex));
        if (!entryEquity.isGreaterThan(barSeries.numFactory().zero())) {
            return;
        }
        int ratioIndex = endIndex;
        if (ratioIndex == entryIndex && entryIndex < barSeries.getEndIndex()) {
            ratioIndex = entryIndex + 1;
        }

        if (equityCurveMode == EquityCurveMode.MARK_TO_MARKET) {
            Num netExitPrice = AnalysisPositionSupport.markToMarket(this, barSeries, position, endIndex,
                    windowStartIndex, windowEndIndex, (index, netPrice, previousPrice) -> buffer.multiply(index,
                            getIntermediateRatio(isLongTrade, netEntryPrice, netPrice)))
                    .netPrice();
            Num ratio = getIntermediateRatio(isLongTrade, netEntryPrice, netExitPrice);
            if (ratioIndex <= windowEndIndex) {
                buffer.multiply(ratioIndex, ratio);
            }
            if (ratioIndex < windowEndIndex) {
                // ratioIndex + 1 must stay representable: skip the empty
                // successor range instead of letting the increment overflow.
                buffer.multiplyRange(ratioIndex + 1, windowEndIndex, ratio);
            }
            return;
        }

        Trade exit = position.getExit();
        if (exit != null && endIndex >= exit.getIndex()) {
            Num holdingCost = position.getHoldingCost(endIndex);
            Num netExitPrice = addCost(exit.getNetPrice(), holdingCost, isLongTrade);
            Num ratio = getIntermediateRatio(isLongTrade, netEntryPrice, netExitPrice);
            buffer.multiplyRange(Math.max(ratioIndex, windowStartIndex), windowEndIndex, ratio);
        }
    }

    /**
     * @param index the bar index
     * @return the cash flow value at the index-th position, or the neutral value
     *         one for indices outside the materialized window
     */
    @Override
    public Num getValue(int index) {
        return values.get(index);
    }

    /**
     * @return values over the captured materialized window, independent of later
     *         changes to the borrowed series bounds
     * @since 0.25.1
     */
    @Override
    public Stream<Num> stream() {
        return values.stream();
    }

    @Override
    public int getCountOfUnstableBars() {
        return 0;
    }

    @Override
    @SuppressFBWarnings(value = "EI_EXPOSE_REP", justification = "Returns the borrowed caller series by contract.")
    public BarSeries getBarSeries() {
        return barSeries;
    }

    /**
     * Returns the first absolute index of the captured curve, independent of later
     * changes to the borrowed series.
     *
     * @return the captured begin index
     * @since 0.25.1
     */
    @Override
    public int getBeginIndex() {
        return window.beginIndex();
    }

    /**
     * {@inheritDoc}
     *
     * @since 0.25.1
     */
    @Override
    public int getEndIndex() {
        return window.endIndex();
    }

    /**
     * @return the number of values captured in the materialized window, unaffected
     *         by later changes to the borrowed series
     */
    public int getSize() {
        return values.size();
    }

    /**
     * @return the equity curve mode used for this cash flow
     * @since 0.22.2
     */
    @Override
    public EquityCurveMode getEquityCurveMode() {
        return equityCurveMode;
    }

    private static Num getIntermediateRatio(boolean isLongTrade, Num entryPrice, Num exitPrice) {
        if (isLongTrade) {
            return exitPrice.dividedBy(entryPrice);
        }
        return entryPrice.getNumFactory().numOf(2).minus(exitPrice.dividedBy(entryPrice));
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/analysis/CashFlow.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.analysis

import org.ta4j.core.BarSeries
import org.ta4j.core.Indicator
import org.ta4j.core.Position
import org.ta4j.core.TradingRecord
import org.ta4j.core.num.*
import java.util.*
import java.util.function.Consumer
import kotlin.math.max
import kotlin.math.min

/**
 * The cash flow.
 *
 * This class allows to follow the money cash flow involved by a list of
 * positions over a bar series.
 */
class CashFlow : Indicator<Num> {
    /**
     * The bar series
     */
    override val barSeries: BarSeries

    /**
     * The cash flow values
     */
    private var values: MutableList<Num>

    /**
     * Constructor for cash flows of a closed position.
     *
     * @param barSeries the bar series
     * @param position  a single position
     */
    constructor(barSeries: BarSeries, position: Position) {
        this.barSeries = barSeries
        values = ArrayList(listOf(numOf(1)))
        calculate(position)
        fillToTheEnd()
    }

    /**
     * Constructor for cash flows of closed positions of a trading record.
     *
     * @param barSeries     the bar series
     * @param tradingRecord the trading record
     */
    constructor(barSeries: BarSeries, tradingRecord: TradingRecord) {
        this.barSeries = barSeries
        values = ArrayList(listOf(numOf(1)))
        calculate(tradingRecord)
        fillToTheEnd()
    }

    /**
     * Constructor.
     *
     * @param barSeries     the bar series
     * @param tradingRecord the trading record
     * @param finalIndex    index up until cash flows of open positions are
     * considered
     */
    constructor(barSeries: BarSeries, tradingRecord: TradingRecord, finalIndex: Int) {
        this.barSeries = barSeries
        values = ArrayList(listOf(numOf(1)))
        calculate(tradingRecord, finalIndex)
        fillToTheEnd()
    }

    /**
     * @param index the bar index
     * @return the cash flow value at the index-th position
     */
    override fun getValue(index: Int): Num {
        return values[index]
    }


    override fun numOf(number: Number): Num {
        return barSeries.numOf(number)
    }

    /**
     * @return the size of the bar series
     */
    val size: Int
        get() = barSeries.barCount

    /**
     * Calculates the cash flow for a single closed position.
     *
     * @param position a single position
     */
    private fun calculate(position: Position?) {
        require(!position!!.isOpened) { "Position is not closed. Final index of observation needs to be provided." }
        calculate(position, position.exit!!.index)
    }

    /**
     * Calculates the cash flow for a single position (including accrued cashflow
     * for open positions).
     *
     * @param position   a single position
     * @param finalIndex index up until cash flow of open positions is considered
     */
    private fun calculate(position: Position, finalIndex: Int) {
        val isLongTrade = position.entry!!.isBuy
        val endIndex = determineEndIndex(position, finalIndex, barSeries.endIndex)
        val entryIndex = position.entry!!.index
        val begin = entryIndex + 1
        if (begin > values.size) {
            val lastValue = values[values.size - 1]
            values.addAll(Collections.nCopies(begin - values.size, lastValue))
        }
        // Trade is not valid if net balance at the entryIndex is negative
        if (values[values.size - 1].isGreaterThan(values[0].numOf(0))) {
            val startingIndex = max(begin, 1)
            val nPeriods = endIndex - entryIndex
            val holdingCost = position.getHoldingCost(endIndex)
            val avgCost = holdingCost.div(holdingCost.numOf(nPeriods))

            // Add intermediate cash flows during position
            val netEntryPrice = position.entry!!.netPrice!!
            for (i in startingIndex until endIndex) {
                val intermediateNetPrice = addCost(barSeries.getBar(i).closePrice!!, avgCost, isLongTrade)
                val ratio = getIntermediateRatio(isLongTrade, netEntryPrice, intermediateNetPrice)
                values.add(values[entryIndex].times(ratio))
            }

            // add net cash flow at exit position
            val exitPrice: Num = if (position.exit != null) {
                position.exit!!.netPrice
            } else {
                barSeries.getBar(endIndex).closePrice
            }!!
            val ratio = getIntermediateRatio(isLongTrade, netEntryPrice, addCost(exitPrice, avgCost, isLongTrade))
            values.add(values[entryIndex].times(ratio))
        }
    }

    /**
     * Calculates the cash flow for the closed positions of a trading record.
     *
     * @param tradingRecord the trading record
     */
    private fun calculate(tradingRecord: TradingRecord) {
        // For each position...
        tradingRecord.positions.forEach(Consumer { position: Position? -> this.calculate(position) })
    }

    /**
     * Calculates the cash flow for all positions of a trading record, including
     * accrued cash flow of an open position.
     *
     * @param tradingRecord the trading record
     * @param finalIndex    index up until cash flows of open positions are
     * considered
     */
    private fun calculate(tradingRecord: TradingRecord, finalIndex: Int) {
        calculate(tradingRecord)

        // Add accrued cash flow of open position
        if (tradingRecord.getCurrentPosition().isOpened) {
            calculate(tradingRecord.getCurrentPosition(), finalIndex)
        }
    }

    /**
     * Fills with last value till the end of the series.
     */
    private fun fillToTheEnd() {
        if (barSeries.endIndex >= values.size) {
            val lastValue = values[values.size - 1]
            values.addAll(Collections.nCopies(barSeries.endIndex - values.size + 1, lastValue))
        }
    }

    companion object {
        /**
         * Calculates the ratio of intermediate prices.
         *
         * @param isLongTrade true, if the entry trade type is BUY
         * @param entryPrice  price ratio denominator
         * @param exitPrice   price ratio numerator
         */
        private fun getIntermediateRatio(isLongTrade: Boolean, entryPrice: Num, exitPrice: Num): Num {
            return if (isLongTrade) {
                exitPrice.div(entryPrice)
            } else {
                entryPrice.numOf(2).minus(exitPrice.div(entryPrice))
            }
        }

        /**
         * Adjusts (intermediate) price to incorporate trading costs.
         *
         * @param rawPrice    the gross asset price
         * @param holdingCost share of the holding cost per period
         * @param isLongTrade true, if the entry trade type is BUY
         */
        fun addCost(rawPrice: Num, holdingCost: Num, isLongTrade: Boolean): Num {
            return if (isLongTrade) {
                rawPrice.minus(holdingCost)
            } else {
                rawPrice.plus(holdingCost)
            }
        }

        /**
         * Determines the the valid final index to be considered.
         *
         * @param position   the position
         * @param finalIndex index up until cash flows of open positions are considered
         * @param maxIndex   maximal valid index
         */
        fun determineEndIndex(position: Position, finalIndex: Int, maxIndex: Int): Int {
            var idx = finalIndex
            // After closing of position, no further accrual necessary
            if (position.exit != null) {
                idx = min(position.exit!!.index, finalIndex)
            }
            // Accrual at most until maximal index of asset data
            if (idx > maxIndex) {
                idx = maxIndex
            }
            return idx
        }
    }
}